[Completion] Handle unbound generics in typealiases

hamishknight opened the pull request:
The underlying type for a typealias can be an unbound generic type, replace these with the bound generic equivalent. This avoids crashing when attempting to compute the type relation (in the future we'll want to open these type parameters for the comparison).

rdar://147789214

hamishknight commented:
@swift-ci please test

slavapestov commented on lib/IDE/CompletionLookup.cpp:
```diff
@@ -1755,6 +1755,42 @@ void CompletionLookup::addNominalTypeRef(const NominalTypeDecl *NTD,
+  // Introduce type parameters such that we print the underlying type as
+  // 'S<T>'.
+  return underlyingTy.transformRec([&](TypeBase *ty) -> std::optional<Type> {
+    if (!ty->hasUnboundGenericType()) {
```
In fact you don't have to recurse with transformRec() because only the outermost type can be unbound. In this case, the type alias behaves like a generic type alias:
```
typealias X<T> = S<T>
```
Can you handle this like a generic type alias?

hamishknight replied:
> In fact you don't have to recurse with transformRec()

Yeah I was just being defensive, I can change to be top-level only

> Can you handle this like a generic type alias?

Would that be any different to `return UGT->getDecl()->getDeclaredInterfaceType()`?

hamishknight replied:
Heh actually you could get a non-top-level UnboundGenericType for an underlying type that is also a typealias, maybe we ought to consider recursively looking through to the underlying type though

slavapestov replied:
I’m not sure what you mean, like this?

```
typealias A = Array
typealias B = A
```

hamishknight replied:
Yeah, we're currently only unwrapping a single level of sugar here so `hasUnboundGenericType` would still be true if we only looked at a top-level UnboundGenericType

slavapestov commented on test/IDE/complete_rdar147789214.swift:
```diff
@@ -0,0 +1,11 @@
+
+struct S<T> {}
+typealias Foo = S
+typealias Invalid = (S, S)
```
The underlying type here should just be an ErrorType

hamishknight replied:
Indeed

slavapestov commented on test/IDE/complete_rdar147789214.swift:
```diff
@@ -0,0 +1,11 @@
+typealias Invalid = (S, S)
+
+let _: S = #^COMPLETE^#
+// COMPLETE-DAG: Decl[TypeAlias]/CurrModule: Foo[#S<T>#]; name=Foo
```
Is this different than what you'd get if you wrote `typealias Foo<T> = S<T>`?

hamishknight replied:
I don't believe so, no

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@swift-ci please smoke test

hamishknight committed:
[Completion] Handle unbound generics in typealiases

The underlying type for a typealias can be an unbound generic type,
replace these with the bound generic equivalent. This avoids crashing
when attempting to compute the type relation (in the future we'll
want to open these type parameters for the comparison).

rdar://147789214

hamishknight commented:
Spoke to Slava offline, updated to remove the use of `transformRec`

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please smoke test macOS
